File: Torry's Delphi Pages/Component Pack/tpessence/examples/CBuildr3/ExMdi1.cpp
// ***** BEGIN LICENSE BLOCK *****
// * Version: MPL 1.1
// *
// * The contents of this file are subject to the Mozilla Public License Version
// * 1.1 (the "License"); you may not use this file except in compliance with
// * the License. You may obtain a copy of the License at
// * http://www.mozilla.org/MPL/
// *
// * Software distributed under the License is distributed on an "AS IS" basis,
// * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
// * for the specific language governing rights and limitations under the
// * License.
// *
// * The Original Code is TurboPower Essentials Vol I
// *
// * The Initial Developer of the Original Code is
// * TurboPower Software
// *
// * Portions created by the Initial Developer are Copyright (C) 1997-2002
// * the Initial Developer. All Rights Reserved.
// *
// * Contributor(s):
// *
// * ***** END LICENSE BLOCK *****

//---------------------------------------------------------------------------
#include <vcl\vcl.h>
#pragma hdrstop

#include "ExMdi1.h"
#include "ExMdi2.h"
//---------------------------------------------------------------------------
#pragma link "EsTile"
#pragma link "EsGrad"
#pragma resource "*.dfm"
TMainForm *MainForm;
//---------------------------------------------------------------------------
__fastcall TMainForm::TMainForm(TComponent* Owner)
  : TForm(Owner)
{
}
//---------------------------------------------------------------------------
void TMainForm::CreateMDIChild(const String Name)
{
  // create a new MDI child window
  TMDIChild* Child = new TMDIChild(Application);
  Child->Caption = Name;
}

void __fastcall TMainForm::ClientWndProc(TMessage& Message)
{
  TRect R;
    switch (Message.Msg) {
      case WM_ERASEBKGND : {
          // erase the background
          CallWindowProc((FARPROC)FPrevClientProc, ClientHandle,
            Message.Msg, Message.WParam, Message.LParam);

          // get the MDI client's rectangle
          GetWindowRect(ClientHandle, (RECT*)&R);

          // Delphi and the Essentials PaintTo methods think that
          // an HDC is an int so we'll use an int here.
          int hDC = (int)Message.WParam;
          if (Tile1->Checked) // draw the tile
            EsTile1->PaintTo(hDC, R);

          if (Gradient1->Checked) // draw the gradient
            EsGradient1->PaintTo(hDC, R);

          Message.Result = 1;
          break;
        }

      case WM_VSCROLL :
      case WM_HSCROLL : {
          Message.Result = CallWindowProc((FARPROC)FPrevClientProc,
            ClientHandle, Message.Msg, Message.WParam, Message.LParam);
          InvalidateRect(ClientHandle, NULL, true);
          break;
      }
      default :
        Message.Result = CallWindowProc((FARPROC)FPrevClientProc,
          ClientHandle, Message.Msg, Message.WParam, Message.LParam);
  }
}

void __fastcall TMainForm::FileNewItemClick(TObject *Sender)
{
  CreateMDIChild("NONAME" + String((int)MDIChildCount + 1));
}
//---------------------------------------------------------------------------
void __fastcall TMainForm::FileOpenItemClick(TObject *Sender)
{
  if (OpenDialog->Execute())
    CreateMDIChild(OpenDialog->FileName);
}
//---------------------------------------------------------------------------
void __fastcall TMainForm::FileCloseItemClick(TObject *Sender)
{
  if (ActiveMDIChild)
    ActiveMDIChild->Close();
}
//---------------------------------------------------------------------------
void __fastcall TMainForm::FileSaveItemClick(TObject *Sender)
{
  // save current file (ActiveMDIChild points to the window)
}
//---------------------------------------------------------------------------
void __fastcall TMainForm::FileSaveAsItemClick(TObject *Sender)
{
  // save current file under new name
}
//---------------------------------------------------------------------------
void __fastcall TMainForm::FileExitItemClick(TObject *Sender)
{
  Close();  
}
//---------------------------------------------------------------------------
void __fastcall TMainForm::CutItemClick(TObject *Sender)
{
  // cut selection to clipboard
}
//---------------------------------------------------------------------------
void __fastcall TMainForm::CopyItemClick(TObject *Sender)
{
  // copy selection to clipboard
}
//---------------------------------------------------------------------------
void __fastcall TMainForm::PasteItemClick(TObject *Sender)
{
  // paste selection from clipboard
}
//---------------------------------------------------------------------------
void __fastcall TMainForm::WindowCascadeItemClick(TObject *Sender)
{
  Cascade();  
}
//---------------------------------------------------------------------------
void __fastcall TMainForm::WindowTileItemClick(TObject *Sender)
{
  Tile();  
}
//---------------------------------------------------------------------------
void __fastcall TMainForm::WindowArrangeItemClick(TObject *Sender)
{
  ArrangeIcons();  
}
//---------------------------------------------------------------------------
void __fastcall TMainForm::WindowMinimizeItemClick(TObject *Sender)
{
  // Must be done backwards through the MDIChildren array
  for (int i=MDIChildCount - 1;i>0;i++)
    MDIChildren[i]->WindowState = wsMinimized;
}
//---------------------------------------------------------------------------
void __fastcall TMainForm::UpdateMenuItems(TObject* Sender)
{
  FileCloseItem->Enabled = MDIChildCount > 0;
  FileSaveItem->Enabled = MDIChildCount > 0;
  FileSaveAsItem->Enabled = MDIChildCount > 0;
  CutItem->Enabled = MDIChildCount > 0;
  CopyItem->Enabled = MDIChildCount > 0;
  PasteItem->Enabled = MDIChildCount > 0;
  SaveBtn->Enabled = MDIChildCount > 0;
  CutBtn->Enabled = MDIChildCount > 0;
  CopyBtn->Enabled = MDIChildCount > 0;
  PasteBtn->Enabled = MDIChildCount > 0;
  WindowCascadeItem->Enabled = MDIChildCount > 0;
  WindowTileItem->Enabled = MDIChildCount > 0;
  WindowArrangeItem->Enabled = MDIChildCount > 0;
  WindowMinimizeItem->Enabled = MDIChildCount > 0;
}

void __fastcall TMainForm::Tile1Click(TObject *Sender)
{
  Tile1->Checked = True;
  Gradient1->Checked = False;
  InvalidateRect(ClientHandle, NULL, true);
}
//---------------------------------------------------------------------------
void __fastcall TMainForm::Gradient1Click(TObject *Sender)
{
  Tile1->Checked = False;
  Gradient1->Checked = True;
  InvalidateRect(ClientHandle, NULL, true);
}
//---------------------------------------------------------------------------
void __fastcall TMainForm::FormCreate(TObject *Sender)
{
  Screen->OnActiveFormChange = UpdateMenuItems;

  // hook the window procedure
  FClientInstance = MakeObjectInstance(ClientWndProc);
  FPrevClientProc = reinterpret_cast<void*>(GetWindowLong(ClientHandle, GWL_WNDPROC));
  SetWindowLong(ClientHandle, GWL_WNDPROC, reinterpret_cast<long>(FClientInstance));
}
//---------------------------------------------------------------------------
void __fastcall TMainForm::FormDestroy(TObject *Sender)
{
  Screen->OnActiveFormChange = 0;

  // restore previous window proc
  SetWindowLong(ClientHandle, GWL_WNDPROC, reinterpret_cast<long>(FPrevClientProc));
  FreeObjectInstance(FClientInstance);
}
//---------------------------------------------------------------------------

File: Torry's Delphi Pages/Component Pack/tpessence/examples/Delphi/ExMdi1.pas

(* ***** BEGIN LICENSE BLOCK *****
 * Version: MPL 1.1
 *
 * The contents of this file are subject to the Mozilla Public License Version
 * 1.1 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * The Original Code is TurboPower Essentials Vol I
 *
 * The Initial Developer of the Original Code is
 * TurboPower Software
 *
 * Portions created by the Initial Developer are Copyright (C) 1997-2002
 * the Initial Developer. All Rights Reserved.
 *
 * Contributor(s):
 *
 * ***** END LICENSE BLOCK ***** *)

unit ExMdi1;

interface

uses
  {$IFDEF Win32} Windows, {$ELSE} WinTypes, WinProcs, {$ENDIF}
  SysUtils, Classes, Graphics, Forms, Controls, Menus,
  StdCtrls, Dialogs, Buttons, Messages, ExtCtrls, EsGrad, EsTile;

type
  TMainForm = class(TForm)
    MainMenu1: TMainMenu;
    File1: TMenuItem;
    FileNewItem: TMenuItem;
    FileOpenItem: TMenuItem;
    FileCloseItem: TMenuItem;
    Window1: TMenuItem;
    N1: TMenuItem;
    FileExitItem: TMenuItem;
    WindowCascadeItem: TMenuItem;
    WindowTileItem: TMenuItem;
    WindowArrangeItem: TMenuItem;
    OpenDialog: TOpenDialog;
    FileSaveItem: TMenuItem;
    FileSaveAsItem: TMenuItem;
    Edit1: TMenuItem;
    CutItem: TMenuItem;
    CopyItem: TMenuItem;
    PasteItem: TMenuItem;
    WindowMinimizeItem: TMenuItem;
    SpeedPanel: TPanel;
    OpenBtn: TSpeedButton;
    SaveBtn: TSpeedButton;
    CutBtn: TSpeedButton;
    CopyBtn: TSpeedButton;
    PasteBtn: TSpeedButton;
    ExitBtn: TSpeedButton;
    EsTile1: TEsTile;
    EsGradient1: TEsGradient;
    Background1: TMenuItem;
    Tile1: TMenuItem;
    Gradient1: TMenuItem;
    procedure FormCreate(Sender: TObject);
    procedure FileNewItemClick(Sender: TObject);
    procedure WindowCascadeItemClick(Sender: TObject);
    procedure UpdateMenuItems(Sender: TObject);
    procedure WindowTileItemClick(Sender: TObject);
    procedure WindowArrangeItemClick(Sender: TObject);
    procedure FileCloseItemClick(Sender: TObject);
    procedure FileOpenItemClick(Sender: TObject);
    procedure FileExitItemClick(Sender: TObject);
    procedure FileSaveItemClick(Sender: TObject);
    procedure FileSaveAsItemClick(Sender: TObject);
    procedure CutItemClick(Sender: TObject);
    procedure CopyItemClick(Sender: TObject);
    procedure PasteItemClick(Sender: TObject);
    procedure WindowMinimizeItemClick(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure Tile1Click(Sender: TObject);
    procedure Gradient1Click(Sender: TObject);
  private
    { Private declarations }

    {!!!variables to hold the address of the old and new window proc!!!}
    FClientInstance : TFarProc;
    FPrevClientProc : TFarProc;

    procedure CreateMDIChild(const Name: string);

    {!!!the new window proc!!!}
    procedure ClientWndProc(var Message : TMessage);
  public
    { Public declarations }
  end;

var
  MainForm: TMainForm;

implementation

{$R *.DFM}

uses
  ExMdi2;

procedure TMainForm.CreateMDIChild(const Name: string);
var
  Child: TMDIChild;
begin
  { create a new MDI child window }
  Child := TMDIChild.Create(Application);
  Child.Caption := Name;
end;

procedure TMainForm.FileNewItemClick(Sender: TObject);
begin
  CreateMDIChild('NONAME' + IntToStr(MDIChildCount + 1));
end;

procedure TMainForm.FileOpenItemClick(Sender: TObject);
begin
  if OpenDialog.Execute then
    CreateMDIChild(OpenDialog.FileName);
end;

procedure TMainForm.FileCloseItemClick(Sender: TObject);
begin
  if ActiveMDIChild <> nil then
    ActiveMDIChild.Close;
end;

procedure TMainForm.FileSaveItemClick(Sender: TObject);
begin
  { save current file (ActiveMDIChild points to the window) }
end;

procedure TMainForm.FileSaveAsItemClick(Sender: TObject);
begin
  { save current file under new name }
end;

procedure TMainForm.FileExitItemClick(Sender: TObject);
begin
  Close;
end;

procedure TMainForm.CutItemClick(Sender: TObject);
begin
  {cut selection to clipboard}
end;

procedure TMainForm.CopyItemClick(Sender: TObject);
begin
  {copy selection to clipboard}
end;

procedure TMainForm.PasteItemClick(Sender: TObject);
begin
  {paste from clipboard}
end;

procedure TMainForm.WindowCascadeItemClick(Sender: TObject);
begin
  Cascade;
end;

procedure TMainForm.WindowTileItemClick(Sender: TObject);
begin
  Tile;
end;

procedure TMainForm.WindowArrangeItemClick(Sender: TObject);
begin
  ArrangeIcons;
end;

procedure TMainForm.WindowMinimizeItemClick(Sender: TObject);
var
  I: Integer;
begin
  { Must be done backwards through the MDIChildren array }
  for I := MDIChildCount - 1 downto 0 do
    MDIChildren[I].WindowState := wsMinimized;
end;

procedure TMainForm.UpdateMenuItems(Sender: TObject);
begin
  FileCloseItem.Enabled := MDIChildCount > 0;
  FileSaveItem.Enabled := MDIChildCount > 0;
  FileSaveAsItem.Enabled := MDIChildCount > 0;
  CutItem.Enabled := MDIChildCount > 0;
  CopyItem.Enabled := MDIChildCount > 0;
  PasteItem.Enabled := MDIChildCount > 0;
  SaveBtn.Enabled := MDIChildCount > 0;
  CutBtn.Enabled := MDIChildCount > 0;
  CopyBtn.Enabled := MDIChildCount > 0;
  PasteBtn.Enabled := MDIChildCount > 0;
  WindowCascadeItem.Enabled := MDIChildCount > 0;
  WindowTileItem.Enabled := MDIChildCount > 0;
  WindowArrangeItem.Enabled := MDIChildCount > 0;
  WindowMinimizeItem.Enabled := MDIChildCount > 0;
end;

{!!!menu event handler to choose a gradient background!!!}
procedure TMainForm.Tile1Click(Sender: TObject);
begin
  Tile1.Checked := True;
  Gradient1.Checked := False;
  InvalidateRect(ClientHandle, nil, True);
end;

{!!!menu event handler to choose a tiled background!!!}
procedure TMainForm.Gradient1Click(Sender: TObject);
begin
  Tile1.Checked := False;
  Gradient1.Checked := True;
  InvalidateRect(ClientHandle, nil, True);
end;

procedure TMainForm.FormCreate(Sender: TObject);
begin
  Screen.OnActiveFormChange := UpdateMenuItems;

  {!!!hook the window procedure!!!}
  FClientInstance := MakeObjectInstance(ClientWndProc);
  FPrevClientProc := Pointer(GetWindowLong(ClientHandle, GWL_WNDPROC));
  SetWindowLong(ClientHandle, GWL_WNDPROC, LongInt(FClientInstance));
end;

procedure TMainForm.FormDestroy(Sender: TObject);
begin
  Screen.OnActiveFormChange := nil;

  {!!!restore previous window proc!!!}
  SetWindowLong(ClientHandle, GWL_WNDPROC, LongInt(FPrevClientProc));
  FreeObjectInstance(FClientInstance);
end;

{!!!new window procedure to trap background erase and scroll messages!!!}
procedure TMainForm.ClientWndProc(var Message : TMessage);
var
  R : TRect;
begin
  with Message do begin
    case Msg of
      WM_ERASEBKGND :
        begin
          {erase the background}
          CallWindowProc(FPrevClientProc, ClientHandle, Msg, wParam, lParam);

          {get the MDI client's rectangle}
          GetWindowRect(ClientHandle, R);

          {usae the PaintTo method to draw the}
          if Tile1.Checked then  {draw the tile}
            EsTile1.PaintTo(TWMEraseBkGnd(Message).DC, R);

          if Gradient1.Checked then  {draw the gradient}
            EsGradient1.PaintTo(TWMEraseBkGnd(Message).DC, R);

          Result := 1;
        end;

      WM_VSCROLL, WM_HSCROLL :
        begin
          Result := CallWindowProc(
            FPrevClientProc, ClientHandle, Msg, wParam, lParam);
          InvalidateRect(ClientHandle, nil, True);
        end;

    else
      Result := CallWindowProc( FPrevClientProc, ClientHandle, Msg, wParam, lParam);
    end;
  end;
end;


end.
